### ISSUE TITLE
Create CSV Compound toolchain

### ISSUE THREAD

**Issue by mjordan** [MarcusBarnes/mik#234]
Initial requirements are to create ingest packages for flat (non-hierarchical) objects, with ooptional child-level metadata. The CSV Newspapers toolchain is a good place to start.

**Comment by mjordan** [comment]
The source files should be arranged like this:

```
├── cpd1
│   ├── image_01.tif
│   └── image_02.tif
├── cpd2
│   ├── image_01.tif
│   └── image_02.tif
└── cpd3
    ├── image_01.tif
    └── image_02.tif
```

The metadata should be structured like this:

```
Identifier,Child,Title,Date,Description
cpd1,,Compound 1,2016-08-05,"I am the first compound object."
cpd2,,Compound 2,2016-08-05,"I am the second compound object."
cpd2,01,"Compound 2 - First image",2016-08-05,"The first image in Compound 2"
cpd2,02,"Compound 2 - Second image",2016-08-05,"The second image in Compound 2"
cpd3,,Compound 3,2016-08-05,"I am the third compound object."
```

The Identifier and Child columns are reserved (although like other CSV toolchains the actual column names are configurable in the .ini file). The `identifier` column indicates the directory name of the compound object, and all other non-reserved fields in the row contain metadata that applies to the compound object. The `Child` column contains the sequence numbers of the child that row applies to, and is only required when there is metadata that should be applied to the child object (if there is no child-level metadata, the Child column must be empty). The child-level metadata uses the same non-reserved columns (Title, Date, etc.) as the rows that describe the compound objects. Rows that contain child sequence numbers must also indicate the child's parent in the Identifier column.

In the example above, only the children of `cpd2` have their own metadata.

**Comment by mjordan** [comment]
Substantially done, but still needs some testing and documentation. Here's a same .ini file:

```
; MIK configuration file for generating Islandora ingest packages
; from a CSV metadata file for flat compound objects.

[CONFIG]
config_id = CSVCompoundTest
last_updated_on = "2016-08-06"
last_update_by = "mjordan@sfu.ca"

[FETCHER]
class = Csv
input_file = '/home/mark/Downloads/mik_csv_compound_metadata.csv'
temp_directory = "/tmp/csv_compound_temp"
record_key = ID
; required
child_key = Child
field_delimiter = '	'

[METADATA_PARSER]
class = mods\CsvToMods
mapping_csv_path = 'csv_compound_test_mappings.csv'
temp_directory = "/tmp/csv_compound_temp"

[FILE_GETTER]
class = CsvCompound
temp_directory = "/tmp/csv_compound_temp"
input_directory = "/home/mark/Downloads/mik_compound_test_data"
; required
cpd_directory_field = Directory

[WRITER]
class = CsvCompound
output_directory = "/tmp/csv_compound_output"
; required
compound_directory_field = Directory
; required
child_title = "%parent_title%, page %sequence_number%"
; child_sequence_separator = '_'
metadata_filename = MODS.xml
; datastreams[] = OBJ
datastreams[] = MODS

[MANIPULATORS]
metadatamanipulators[] = "FilterModsTopic|subject"

[LOGGING]
path_to_log = "/tmp/csv_compound_output/mik.log"
path_to_manipulator_log = "/tmp/csv_compound_output/mik_manipulator.log"
```

Corresponding to this input CSV data:

```
ID	Child	Directory	Title	Date	Description
1		MSC130-08857	The Something Glacier, BC	1909	"Vertical orientation."
2		MSC130-08863	Mt. Sir MacDonald	1912	"Some trees in the foreground."
3		MSC130-08865	Mt. Sir MacDonald Glacier	1918	"Warning: contains depictions of clouds."
4	02	MSC130-08865	Back of postcard		"Has 'xii' written on back"
```

**Comment by mjordan** [comment]
Draft wiki page at https://github.com/MarcusBarnes/mik/wiki/Toolchain:-CSV-compound-objects

**Comment by MarcusBarnes** [comment]
Closed with commit 7b7e0d24b89ce8b795c539c4932a56759f823754.  Thank you @mjordan for this nice toolchain.
